1747 grade routes edit refactor

jwright opened the pull request:
Moves the `#edit` and `#update` endpoints for a `GradesController` to use a `grade_id` instead of an `assignment_id` and `student_id`. This cleaned up the controller methods quite a bit and used resourceful routing for the two actions.

For rubric grades, an update was necessary for the Angular code so it no longer gathered the `assignment_id` and `student_id` or `group_id` from the url. Instead, these are now passed in via `ng-init` to the front end from the erb view.

In addition, the old grade update path performed a `Grade.find_or_create` in the action. This is not possible in the new `Grade#edit` path because a grade id is required. So the path `POST /assignments/:assignment_id/students/:student_id/grade` path was created to perform a `Grade.find_or_create` and then redirect to `Grade#edit` when the grade is not known.

chcholman commented on app/controllers/assignments/groups_controller.rb:
```diff
@@ -9,12 +9,7 @@ def grade
-    if @assignment.grade_with_rubric?
-      @rubric = @assignment.rubric
-      # This is sent to the Angular controlled submit button
-      @return_path = URI(request.referer).path
```
I'm concerned about losing the redirection here - is this being handled in a different way?

chcholman commented on app/controllers/grades_controller.rb:
```diff
@@ -18,52 +17,34 @@ def show
-      @criterion_grades = serialized_criterion_grades
-      # This is sent to the Angular controlled submit button
-      @return_path =
-        URI(request.referer).path + "?student_id=#{current_student.id}"
```
Same note here as above.

jwright commented on app/views/assignments/individual/_table_body.html.haml:
```diff
@@ -66,15 +66,15 @@
               %li= link_to glyph("unlock-alt") + "Unlock", manually_unlock_unlock_state_path(student_id: student.id, assignment_id: presenter.assignment.id), :method => :post, class: "button"
             - has_submission = presenter.assignment.accepts_submissions? && !submissions_for_assignment.empty?
-            %li= link_to glyph(:check) + "Grade", edit_assignment_grade_path(presenter.assignment, student_id: student.id), class: "button #{has_submission ? "danger" : ""}"
+            %li= link_to glyph(:check) + "Grade", edit_grade_path(grade), class: "button #{has_submission ? "danger" : ""}"
```
@acousticrobot  I believe that `app/views/assignments/group/_table_body.html.haml:59` will never be a `NullGrade` since it's for a group which means that the grade will always have a grade. There is also a check to ensure that the grade is not nil. Why does `NullGrade#is_graded?` return a `true` value?

jwright commented:
@acousticrobot Thank you for your review.

1. I removed the `RubricService#getAssignment` by passing in the `assignmentId` like you suggested. Awesome!
2. I replaced `edit_grade_path` with `assignment_student_grade_path` which does a `Grade.find_or_create` before redirecting to `edit_grade_path` to ensure a grade exists for a few other places than the one you suggested. I replaced where I did not know for certain that a grade would exist. I only kept `edit_grade_path` where I knew a grade would not exist (like it is coming from the database, etc). I am not sure this is the best approach. Perhaps we should just use `assignment_student_grade_path` everywhere?

I am passing this back to you for one more look to ensure that the last commits are good for you. If you are good, it can probably be passed onto cait.
